### DISCUSSION TITLE
Développement module externe

### DISCUSSION THREAD

**Discussion by coder-pour-changer-de-vie** [PnX-SI/GeoNature#2797] [Q&A] [answered]
Bonjour à tous, je suis nouveau dans l'aventure GeoNature, heureux de faire connaissance au fil de l'eau !

Je lis actuellement la documentation en profondeur, voici mon contexte avant de poser ma question : 

1) je débute en "GeoNature"
2) je configure actuellement mon poste de développement pour développer un module externe
3) je travaille sur la version 2.11.0 (car celle en production, il n'est pas prévu de migrer vers la dernière version stable de Geonature prochainement)

De ce que je comprends actuellement, je peux donc passer par une installation de GeoNature sur mon poste de développement en prendant la version 2.11.0 afin de m'en servir comme base de déploiement (je viens d'installer un Debian 11 sur un VM pour cela).

Le developpement doit être réalisé à partir du template de module externe présent ici : https://github.com/PnX-SI/gn_module_template

Je devrais créer un nouveau schema Postgresql spécifique au module externe développé

Ma question : pour vous, quel est le flux de travail le plus efficace en tant que développeur ?

Dois-je partir du template, positionner mon code et ensuite le déployer sur la version locale de GeoNature ?

Si oui, comment voir les changement en direct ? Faut il scripter un déploiement automatique local pour voir les changements ?

Auriez-vous un exemple du process de développement dans ce cadre ou bien quelqu'un avec qui je puisse échanger pour faciliter cette prise en main ?

D'avance merci pour votre aide !

**Comment by camillemonchicourt** [comment]
La méthode conseillée serait : 
- Vous récupérez le template de module
- Vous faites un nouveau dépôt Git pour votre module et vous développez dans ce dépôt Git à part dédié à votre module
- C'est le cas pour tous les modules externes comme https://github.com/PnX-SI/gn_module_export

**Reply by coder-pour-changer-de-vie** [reply]
Bonjour et merci.

Toujours en découverte de l'écosystème GeoNature, j'ai pu avancer sur plusieurs points de configuration : 

- Serveurs Front et Back opérationnels en mode dev dans une VM debian 11 avec la version 2.11.0 (car les version docker sont accessibles dans les version ulterieures de GeoNature, donc j'utilise une VM pour fonctionner en version 2.11.0).

- je suis parti ensuite d'une copie du repo https://github.com/PnX-SI/gn_module_template en modifiant les références à "template" par le nom du module lorsque nécessaire, en me basant sur le parcours du code ainsi que plusieurs modules GeoNature existants (dashboard, import, export).

Malheureusement l'import du module fonctionne en partie, il est ok pour le back, en erreur pour le front et n'a pas d'effet sur le shema sql attendu pour le module (il n'est pas créé). 

Mais pour ce dernier je pense que l'étape doit s'exécuter après celle du front, ce qui expliquerait cela.

Je note quelques différences entre les modules existants (sur github) et le code du template de module et j'ai plusieurs questions (par exemple : faut il créer le fichier gnModule.module.ts "à la main" en se basant sur le fichier app.module.ts, pourquoi n'y a t'il pas de repertoire "src" dans le repo des modules existants, etc...)

ça vient sûrement de moi, j'ai du rater quelques trucs ou peut être y a t'il quelques delta avec la documentation.

C'est donc le moment de proposer mon aide :-)

Quel est le processus à suivre pour participer à la mise à  jour de la documentation du template (https://github.com/PnX-SI/gn_module_template) ?

J'ai toute l'expérience d'un novice pour ajouter les étapes qui m'ont manquées et qui pourront aider les nouveaux participants dans le développement de module :-)

Si quelqu'un peut m'aider un peu sur le sujet ce serait top, je pourrais détailler la problématique de déploiement que je rencontre (qui ne doit pas être grand chose à résoudre).

Et cela au niveau du repos dédié plutôt qu'ici j'imagine.

D'avance merci et bonne journée !

**Reply by jacquesfize** [reply, marked as answer]
Le template présent dans https://github.com/PnX-SI/gn_module_template n'est pas encore à jour. Vous avez deux solutions.

La première est de regarder conjointement la [documentation](https://docs.geonature.fr/development.html#developper-un-module-externe) et un projet de module existant. Pour ce dernier, nous conseillons de regarder le [module d'import](https://github.com/PnX-SI/gn_module_import) qui est le plus récent.

La deuxième solution (encore en travaux) est de vous tourner la branche [cookiecutter-generation](https://github.com/PnX-SI/gn_module_template/tree/cookiecutter-generation). Elle permet d'utiliser `cookie-cutter` qui permet de générer un projet à partir d'un template.

**Reply by coder-pour-changer-de-vie** [reply]
Merci beaucoup ! C'est parti je m'y mets ;-)
